File: medium/133-clone-graph.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;

    public Node() {}

    public Node(int _val,List<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/
class Solution {
    public Node cloneGraph(Node node) {
        Map<Node, Node> map = new HashMap<>();
        return cloneMap(node, map);
    }
    
    private Node cloneMap(Node node, Map<Node, Node> map) {
        Node newNode = new Node(node.val, new LinkedList<>());
        map.put(node, newNode);
        for (Node neighbor : node.neighbors) {
            if (!map.containsKey(neighbor)) {
                cloneMap(neighbor, map);
            }
            newNode.neighbors.add(map.get(neighbor));
        }
        return newNode;
    }
}

File: medium/133-clone-graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = []):
        self.val = val
        self.neighbors = neighbors
"""
class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        if node == None:
            return node
        newNode = Node(node.val, [])
        queue = []
        queue.append(node)
        mapping = {node : newNode}
        while queue:
            curNode = queue.pop(0)
            for neighbor in curNode.neighbors:
                if neighbor not in mapping:
                    mapping[neighbor] = Node(neighbor.val, [])
                    queue.append(neighbor)
                mapping[curNode].neighbors.append(mapping[neighbor])
        return mapping[node]
    
"""
Time Complexity: O(N)
Space O(N)
"""
